[fx]Allow rewrite a symbolic traced module

842974287 opened the pull request:
Summary: After symbolic tracing, `fn` seems to already have "forward" in its globals. In this case, `new_keys` would have length of 0 and we take "forward" from `global_dict` directly as `fn_compiled`.

Test Plan: Added a new test in test_fx_experimental.

Differential Revision: D27049012

842974287 commented on torch/fx/experimental/rewriter.py:
```diff
@@ -38,8 +38,11 @@ def rewrite(self, fn: FunctionType):
-        fn_compiled = globals_dict[new_keys[0]]
+        assert len(new_keys) <= 1
+        if len(new_keys) == 0:
+            fn_compiled = globals_dict["forward"]
```
I just encountered another case where it's not "forward" but "_forward_unimplemented". I'm wondering if it makes sense to make it like 
```
assert len(new_keys) <= 1
fn_compiled = globals_dict[fn.__name__]
```

ansley replied:
We always symbolically trace the function called `forward` (refer to [this](https://github.com/pytorch/pytorch/blob/master/torch/fx/symbolic_trace.py)), so I don't think it makes sense to allow for a non-forward function to be used with the subgraph rewriter. It looks like [_forward_unimplemented](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L190) is an intermediate state for the `forward` function. How are you getting to the point where you're rewriting `_forward_unimplemented`?

842974287 replied:
I see. I just found out that I was seeing "_forward_unimplemented` because I was using acc rewriter which is slightly different from this one and not related to this PR. Sorry for the confusion.

ansley commented on test/test_fx_experimental.py:
```diff
@@ -584,6 +584,21 @@ def test_conv_bn_fusion(self):
 
         self.assertEqual(fused(inp), rn18(inp))
 
+    def test_rewrite_traced_module(self):
```
What new code path is this testing?

842974287 replied:
This is testing tracing a GraphModule with RewritingTracer.

ansley replied:
I might be wrong, but I thought this was already covered by the `test_call_to_assert_...` tests in this file

842974287 replied:
Oh this test case is testing the edge case where we symbolic trace a module first then use RewritingTracer to trace this already traced module while `test_call_to_assert_...` are all calling RewritingTracer on nn.Module. The code path this test case go through is this branch
```
if len(new_keys) == 0:
    fn_compiled = globals_dict["forward"]
```

ansley replied:
I see, sorry I missed that!

842974287 replied:
Totally fine, thanks for reviewing it! Let me know if any other changes are needed.

jamesr66a reviewed (commented):
`GraphModule`'s generated `forward` shouldn't have itself in its `__globals__` dict. This is an artifact of the fact that we're `exec`ing forward in a global scope in `_forward_from_src`. Instead let's use the following change to fix this:

```
diff --git a/torch/fx/graph_module.py b/torch/fx/graph_module.py
index b6533f02d43..b0a3230dd4c 100644
--- a/torch/fx/graph_module.py
+++ b/torch/fx/graph_module.py
@@ -41,8 +41,9 @@ def _forward_from_src(src: str, globals: Dict[str, Any]):
     # avoid mutating the passed in dict
     globals = globals.copy()
     exec_with_source(src, globals)
-    return globals['forward']
-
+    forward_fn = globals['forward']
+    del globals['forward']
+    return forward_fn
 
 def _format_import_statement(name: str, obj: Any, importer: Importer) -> str:
     if name in _custom_builtins:
```

Thanks!

842974287 commented:
Address @jamesr66a's comments

ansley commented on torch/fx/graph_module.py:
```diff
@@ -37,11 +37,14 @@ def patched_getline(*args, **kwargs):
     exec_with_source(src, globals)
-    return globals['forward']
+    forward_fn = globals['forward']
+    del globals['forward']
```
You can delete this line since you're making a local copy of `globals`

ansley replied:
You can delete this line since you're making a local copy of `globals`

ansley commented on torch/fx/graph_module.py:
```diff
@@ -37,11 +37,14 @@ def patched_getline(*args, **kwargs):
     exec_with_source(src, globals)
-    return globals['forward']
+    forward_fn = globals['forward']
+    del globals['forward']
```
You can delete this line since you're making a local copy of `globals`

842974287 replied:
I thought we want to delete 'forward' from `globals` here so that `GraphModule`'s generated `forward` won't have itself in its `__globals__` dict right? And dict copy is doing shallow copy so we would still have to delete `forward` here.

ansley commented on torch/fx/graph_module.py:
```diff
@@ -37,11 +37,14 @@ def patched_getline(*args, **kwargs):
+
 def _forward_from_src(src: str, globals: Dict[str, Any]):
     # avoid mutating the passed in dict
     globals = globals.copy()
```
I would prefer that this be named something like `d` or `globals_copy` instead. I had to think for a second about Python scoping rules, pass-by semantics, and whether the dict `copy` function makes a shallow vs. deep copy to ensure that this line was correct

842974287 replied:
Changed it to `globals_copy`.

842974287 committed:
[fx]Allow rewrite a symbolic traced module (#54011)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54011

After symbolic tracing, `fn` seems to already have "forward" in its globals. In this case, `new_keys` would have length of 0 and we take "forward" from `global_dict` directly as `fn_compiled`.

Test Plan: Added a new test in test_fx_experimental.

Reviewed By: ansley

Differential Revision: D27049012

fbshipit-source-id: 0360f2719e240b8c986c22f2a143d8c7bd8e84ad
